### ISSUE TITLE
Investigate support for Universal/Isomorphic rendering

### ISSUE THREAD

**Issue by addyosmani** [insin/react-hn#27]
In the [application shell architecture](https://medium.com/google-developers/instant-loading-web-apps-with-an-application-shell-architecture-7c0c2f10c73#.tlax37k1e), we try to encourage server-side rendering for fast first paint as much as possible. Even if just for the "shell"/common user interface.

In our current version of react-hn in master, we still rely on client-side rendering to display any content. We can probably do way better here so a static render gives us something useful on the screen. 

I would love to see what we can do to explore [Universal rendering](https://24ways.org/2015/universal-react/) support and what perf gains this would bring. We would want this to play as nicely as possible with our Service Worker caching but..good to get a start on a PR to see what a V0 take on this would look like.

cc @jackfranklin @garbee @zenorocha who may be interested in helping with this or have a few pointers :cake:

**Comment by jackfranklin** [comment]
This is exciting! The first step would be to take https://github.com/insin/react-hn/blob/master/src/index.js and produce a server variant that could run via something like Express. I've no idea how that would play with SWs but I'd be interested to find out!

**Comment by jackfranklin** [comment]
I'm happy to take a stab at it btw if you'd like me to.

**Comment by addyosmani** [comment]
@jackfranklin That would be really appreciated. Would love to see how you think this should be structured.

**Comment by jackfranklin** [comment]
@addyosmani so I started playing around with this and hit the fact that all the stores that are used here reference `window` a lot for local storage and for binding to events. I'm still trying to get my head around all the code here but have you any thoughts on this? Maybe we can just avoid using the stores all together on the server and just render the app shell with no data from the server? That would be a good first step at least?

The other thing to do is polyfill it in with JSDOM or something but I'm not sure I'm too keen on that idea. LMK what you think.

**Comment by jackfranklin** [comment]
So for now I wrapped every `window` or `document` reference in an `if typeof ...` check, just to see if we can get it going. When running the server though the first server side render does sort of work, but then I get:

```
Invariant Violation: React DOM tree root should always have a node reference.
    at invariant (/Users/jackfranklin/git/react-hn/node_modules/fbjs/lib/invariant.js:38:15)
    at getNodeFromInstance (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactDOMComponentTree.js:164:67)
    at ReactDOMComponent.Mixin._updateDOMProperties (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactDOMComponent.js:786:20)
    at ReactDOMComponent.Mixin.updateComponent (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactDOMComponent.js:686:10)
    at ReactDOMComponent.Mixin.receiveComponent (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactDOMComponent.js:643:10)
    at ReactDOMComponent.wrapper [as receiveComponent] (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactPerf.js:66:21)
    at Object.ReactReconciler.receiveComponent (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactReconciler.js:103:22)
    at [object Object].ReactCompositeComponentMixin._updateRenderedComponent (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactCompositeComponent.js:653:23)
    at [object Object].ReactCompositeComponentMixin._performComponentUpdate (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactCompositeComponent.js:635:10)
    at [object Object].ReactCompositeComponentMixin.updateComponent (/Users/jackfranklin/git/react-hn/node_modules/react/lib/ReactCompositeComponent.js:564:12)
```

This seems related to https://github.com/facebook/react/issues/6232, but I haven't had time to dig into it yet to figure out what the issue is.

If you want to take my (incredibly hacky, not proper code) and play with it, it's on this branch: https://github.com/jackfranklin/react-hn/tree/universal-experiment.

**Comment by addyosmani** [comment]
@jackfranklin Thanks a lot for exploring this so quickly! I'll give your branch a whirl.

From the looks of that issue, the exception you're seeing may be specific to certain versions of browsers being tested in (but I think it may actually be a React bug). What were you seeing this bug in?

> Maybe we can just avoid using the stores all together on the server and just render the app shell with no data from the server? That would be a good first step at least?

I think that if we need to rethink a little of our architecture for the content side, it would be awesome to at least be able to deliver the app shell without data from the server as a very first step. Especially if it isn't gated on the issue you're seeing after wrapping those refs :)

**Comment by jackfranklin** [comment]
@addyosmani I'm seeing that error on the Node server, not on the browser. The server returns back some HTML (which looks fine) but then crashes with that error. Not entirely sure what's going on there.

Agree that 1st step is delivering the shell, I don't see any reason why that isn't possible with the refs wrapped :)

**Comment by addyosmani** [comment]
@jackfranklin It looks like that error you were running into is reproducible with React 15. I haven't been able to find a workaround that worked for me in https://github.com/facebook/react/issues/6232, but I might also be missing something totally obvious.

**Comment by psimyn** [comment]
@jackfranklin I got your branch working by changing `componentWillMount` to `componentDidMount` in `Stories` component. Haven't gone through it to find what is breaking on server, but I suspect something in StoryStore.

**Comment by addyosmani** [comment]
@psimyn Were you able to get the shell fully rendering in browser when you switched to componentWillMount? We do currently use sessionStorage and localStorage for caching history positions and a very high level set of item cache indices, but I think it's possible to work around this.

**Comment by psimyn** [comment]
Seems to all work fine - nav/paging links all work (loading page with JS disabled).

![shell](https://cloud.githubusercontent.com/assets/1731922/15071394/878630c0-13ce-11e6-85d9-c7e84e098fca.png)

I think protection around the cache setting is needed - `onStoriesUpdated` is where it seems to break when using `componentWillMount`

**Comment by addyosmani** [comment]
I think you're right. Anyone interested in playing with shielding caching as a next step here?

I've been trying to play with/figure out what the major blockers for server rendering the main content (topstories, comments) are. Beyond shields, we currently use the main real-time Firebase API for content. I wonder if we would need to switch to (or fallback) to their REST API and use node-fetch + just JSON responses to get the data working as expected. 

I'm also looking at https://github.com/jsdf/hacker-news-mobile, https://github.com/camsong/redux-hacker-news and https://github.com/moretti/react-isomorphic-hn for pointers.

(Side: if we do get isomorphic rendering working it will help a lot with the offline support I've been exploring)

**Comment by addyosmani** [comment]
cc @davideast just as an FYI

**Comment by addyosmani** [comment]
For anyone interested: https://github.com/insin/react-hn/tree/isomorphic-take1 includes @jackfranklin's work, some further tweaks I made to bring back hostname parsing without using the DOM and one or two of the fixes mentioned in this thread to fix breakage.

**Comment by addyosmani** [comment]
So we've now landed universal rendering for the UI shell. We don't yet do this for the content. As part of introducing 'Offline Mode' (in Settings) we landed support for a REST-based HN Service over in https://github.com/insin/react-hn/blob/master/src/services/HNServiceRest.js. 

@jackfranklin suggested that we could maybe use https://github.com/ericclemmons/react-resolver as way to get universal rendering working for us, maybe using the REST service. Perhaps we could expose this initially as a Setting that would let you switch on full Universal rendering support? i.e if on, use resolver + REST instead of live updates.

**Comment by jackfranklin** [comment]
I suggest react-resolver only because I've used it once on a small demo and it worked really nicely - happy to take other ideas if anyone else has more experience.

@addyosmani how are you setting that flag ? Are you thinking query param? So default is effectively what we have now, and you go for `?rest=1` or something that would get you REST API + server resolved data and no/some live updates on the client?

**Comment by addyosmani** [comment]
@jackfranklin There are a few ways we could tackle this. After discussing it a little further on Twitter, a few ideas that dropped out:

- By default, use a server-side/universal render (with REST + react-resolver) for the first view of any page. Once JS has loaded, switch to the normal HNService.js adapter (Firebase/WebSockets) for any real-time updates. This avoids needing to introduce any flags or query params
- Consider react-resolver an experiment here. We would trigger it by passing a query-param which would give you REST API + server resolved data and (like above) once JS is fully loaded try to switch to the HNService normal client for updates. A first pass could ignore trying to do the switch.

**Comment by addyosmani** [comment]
I started roughly [playing](https://github.com/insin/react-hn/commit/4b405073f0515772bd6c965fccea6f614cdb3085) with `react-resolver` over the weekend and run into a few brick walls (likely due to my own ignorance).  The server.js changes I made look a little like this:

```js
app.get('*', function(req, res) {
  ReactRouter.match({
    routes: routes,
    location: req.url
  }, function(err, redirectLocation, renderProps) {
    if (err) {
      res.status(500).send(err.message)
    }
    else if (redirectLocation) {
      res.redirect(302, redirectLocation.pathname + redirectLocation.search)
    }
    else if (renderProps) {
      Resolver
        .resolve(function() {return RouterContext(renderProps)})
        .then(function(resolverRes) {
          var markup = renderToString(
            React.createElement(resolverRes.Resolved, renderProps, null)
          )
          res.render('index', { 
            markup: markup,
            scriptTag: '<script>window.__REACT_RESOLVER_PAYLOAD__ = ' + JSON.stringify(resolverRes.data) + '</script>'
          })
      })
    }
    else {
      res.sendStatus(404)
    }
  })
})
```

which I think look rights. **Note that after the next sets of changes I still don't manage to get `resolverRest.data` populated with anything**

I then updated the app's `index.js` to use resolver for rendering instead of the straight-up React.render directly:

```js
var Resolver = require('react-resolver').Resolver

Resolver.render(function() { return <Router history={history} routes={routes}/> }, document.getElementById('app'))
```

I then started looking at what needed to be done to switch one component (Comments) to use resolver for resolving our data. Most components requiring firebase have a `bindFirebaseRef` method called during `componentWillMount`. Rather than calling this, I took inspiration from [isomorphic-demo](https://github.com/allenkim67/isomorphic-demo/blob/2be59306196e84f66041dc7a03bbd0c805d371aa/components/About.js#L21) by @allenkim67 and switched this over to doing the following in the `module.exports` of that component:

```js
module.exports = resolve('comment', function(props) {
  return HNServiceRest.itemRef(props.id).then(function(res) {
    return res.json()
  }).then(function(snapshot) {
    return snapshot
  })
})(Comment)
```

Client-side, snapshots get fetched correctly here (at least, we can log them to console) but nothing appears to happen server-side. We're also no longer using [replaceState](https://github.com/insin/react-hn/blob/4b405073f0515772bd6c965fccea6f614cdb3085/src/Comment.js#L96), which was previously relied on to get the view updated correctly. This might be part of what I'm missing. 

I'll try taking another look at this in a few days. 

Side: I don't feel like I have a solid enough grasp of why none of our network requests for data work server-side without using react-resolver, given that we rely on isomorphic-fetch. I understand resolver gives us async rendering but I probably just need to study this whole topic a little more wrt React :)

**Comment by addyosmani** [comment]
Takeaway from my `react-resolver` experiments: didn't quite work out well for our needs. If someone wants to take a fresh stab at it, more than welcome :)

Until then, I started looking into just using `statics` inside our components to define a `fetchData` routine that could be called from either the client or the server. I've seen a [few](https://github.com/pwmckenna/tally-master/blob/5d07c78bddc4407125c9bd5902317a875c304a2a/app/Scoreboard.jsx#L13) [different](https://jonassebastianohlsson.com/blog/2015/03/24/isomorphic-react-in-real-life/) universal React apps do this and I think it might help us reach our end goal for content rendering. 

Basically, something like:

```js
  statics: {
    fetchData: function(props) {
        return HNServiceRest.itemRef(props.id).then(function(res) {
          return res.json()
        })
    }
  },
```

We would then need to schedule these fetches from our server-side before rendering. Something like  [this](https://github.com/pwmckenna/tally-master/blob/5d07c78bddc4407125c9bd5902317a875c304a2a/app.jsx#L31).

**Comment by addyosmani** [comment]
Updating this thread with progress on local experiments:

* Tried fetchData statics for each component that we `Promise.all` resolve before server rendering
* Tried "Fat" router that fetches the top stories then resolves the child entries before sending back to props

Unfortunately, because of the way the official Firebase HN API is setup, there are quite a lot of network requests that need to be resolved (fetch index of stories, then fetch each story in order to display 'top stories') before we can return anything useful from the server - especially in REST mode. This negates the benefit of SSR in our case because a user ends up waiting up to a minute for content to get returned from the 'local' server. We can imagine similar problems being experienced with the comments page. 

We could work around this by maintaining our own static cache that's periodically refreshed and perform an SSR render from there. I'm going to spend a little time rethinking how we can tackle this problem. The bottleneck atm is less so "can we SSR our content with React" and more so "how do we do this efficiently on the server".
